### PR TITLE
acrnboot: correct the calculation of the end boundry of _DYNAMIC region

### PR DESCRIPTION
**Pull request by bigdata-memory**
The calculation of the end boundry address is corrected by
    adding the size extracted from _DYNAMIC to start address
    in type of uint8_t while improving the code by calulating
    the end boundry address after scanning, also reducing type
    casts accordingly.

Tracked-On: projectacrn#4191
Signed-off-by: Gary <gordon.king@intel.com>

### CODE REVIEW

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

File: hypervisor/boot/reloc.c [hunk 1]
```diff
@@ -73,10 +75,10 @@ void relocate(void)
 		for (dyn = (struct Elf64_Dyn *)_DYNAMIC; dyn->d_tag != DT_NULL; dyn++) {
 			switch (dyn->d_tag) {
 			case DT_RELA:
-				start = (struct Elf64_Rel *)(dyn->d_ptr + delta);
+				rela_start = (uint8_t *)(dyn->d_ptr + delta);
```
Comment (ZideChen0) [inline hypervisor/boot/reloc.c]: Seems we don't need the new variable rela_start. Keep things simple. How about something like this:

@@ -95,7 +95,7 @@ void relocate(void)
                primary_32_start = (uint64_t)(&cpu_primary_start_32) - delta;
                primary_32_end = (uint64_t)(&cpu_primary_start_64) - delta;
 
-               while (start < end) {
+               while (rela_size >= size) {
                        if ((elf64_r_type(start->r_info)) == R_X86_64_RELATIVE) {
                                addr = (uint64_t *)(delta + start->r_offset);
 
@@ -120,6 +120,7 @@ void relocate(void)
                                }
                        }
                        start = (struct Elf64_Rel *)((char *)start + size);
+                       rela_size -= size;
                }
        }

BTW, with the new variable rela_size, to make it clearer,  probably we can change 'size' to 'entry_size'?

File: hypervisor/boot/reloc.c [hunk 2]
```diff
@@ -59,7 +59,9 @@ void relocate(void)
 {
 #ifdef CONFIG_RELOC
 	struct Elf64_Dyn *dyn;
-	struct Elf64_Rel *start = NULL, *end = NULL;
+	struct Elf64_Rel *entry = NULL;
+        uint8_t *rela_start = NULL, *rela_end = NULL;
```
Comment (dongyaozu) [inline hypervisor/boot/reloc.c]: Are u using TAB or SPACE?

Comment (ZideChen0) [reply hypervisor/boot/reloc.c]: There are some spaces in this patch. Need to change to tabs.

Comment (bigdata-memory) [reply hypervisor/boot/reloc.c]: Yes, there are some white spaces introduced inadvertently, I will fix it, thanks you @dongyaozu and @ZideChen0 point it out.

File: hypervisor/boot/reloc.c (hunk 1 above)
Comment (bigdata-memory) [reply hypervisor/boot/reloc.c]: Okay, I will change the 'size' to ' entry_size', Thanks

**Comment by ClaudZhang1995** [comment]
OK to verify

**Comment by HuiHuangShi** [comment]
8 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/boot/reloc.c func:relocate offset:78<br> reason:Cast from integral type to pointer. : (unsigned long to unsigned char*): ( uint8_t * ) ( dyn -> d_ptr + delta )<br><2> config:sharing_config filename:/hypervisor/boot/reloc.c func:relocate offset:100<br> reason:Pointer assigned to NULL may be dereferenced. : rela_start<br><3> config:part_config filename:/hypervisor/boot/reloc.c func:relocate offset:100<br> reason:Pointer assigned to NULL may be dereferenced. : rela_start<br><4> config:common filename:/hypervisor/boot/reloc.c func:relocate offset:101<br> reason:< > <= >= used on different object pointers.<br><5> config:common filename:/hypervisor/boot/reloc.c func:relocate offset:102<br> reason:Cast to an unrelated type. : (unsigned char* to struct Elf64_Rel*): ( struct Elf64_Rel * ) rela_start<br><6> config:common filename:/hypervisor/boot/reloc.c func:relocate offset:102<br> reason:Casting operation to a pointer. : (unsigned char* to struct Elf64_Rel*): ( struct Elf64_Rel * ) rela_start<br><7> config:common filename:/hypervisor/boot/reloc.c func:relocate offset:102<br> reason:Casting operation on a pointer. : (unsigned char* to struct Elf64_Rel*): ( struct Elf64_Rel * ) rela_start<br><8> config:common filename:/hypervisor/boot/reloc.c func:relocate offset:104<br> reason:Cast from integral type to pointer. : (unsigned long to unsigned long*): ( uint64_t * ) ( delta + entry -> r_offset )<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by dongyaozu** [comment]
@ZideChen0 please review

**Comment by ZideChen0** [comment]
Reviewed-by: Zide Chen <zide.chen@intel.com>

**Pull request merged**